File: product-processor-service/src/main/java/com/example/product_processor_service/service/common/kafka/producer/KafkaProducerServiceManager.java
package com.example.product_processor_service.service.common.kafka.producer;

import com.example.product_processor_service.service.common.kafka.template.router.KafkaTemplateRouter;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.io.Serializable;

@Service
public class KafkaProducerServiceManager implements KafkaProducerService {
    private final KafkaTemplateRouter kafkaTemplateRouter;

    public KafkaProducerServiceManager(KafkaTemplateRouter kafkaTemplateRouter) {
        this.kafkaTemplateRouter = kafkaTemplateRouter;
    }

    @Override
    public void send(ProducerRecord<String, ? extends Serializable> producerRecord) {
        sendWithTemplate(kafkaTemplateRouter.route(producerRecord.value().getClass()), producerRecord);
    }

    private <V extends Serializable> void sendWithTemplate(
            KafkaTemplate<String, V> template,
            ProducerRecord<String, ? extends Serializable> record) {
        template.send(record.topic(), (V) record.value());
    }
}

File: product-manager-service/src/main/kotlin/com/example/product_manager_service/service/kafka/producter/KafkaProducerServiceManager.kt
package com.example.product_manager_service.service.kafka.producter

import com.example.product_manager_service.service.kafka.template.router.KafkaTemplateRouter
import org.apache.kafka.clients.producer.ProducerRecord
import org.springframework.stereotype.Service
import java.io.Serializable
import kotlin.reflect.KClass

@Service
class KafkaProducerServiceManager(private val kafkaTemplateRouter: KafkaTemplateRouter) :
    KafkaProducerService {
    override fun send(producerRecord: ProducerRecord<String, out Serializable>) {
        kafkaTemplateRouter
            .route(producerRecord.value()::class as KClass<Serializable>)
            .send(producerRecord.topic(),producerRecord.value())
    }
}
